notebooks/py/DARTFISH.py

```diff
@@ -26,76 +28,89 @@
 # EPY: END code
 
 # EPY: START markdown
-#### Load image stack
+### Load data into Starfish from the Cloud
 #
-#Note that the data here corresopond to DARTFISHv1 2017. The group is actively working on improving the protocol.
+#The example data here corresopond to DARTFISHv1 2017. The group is actively working on improving the protocol. The data represent human brain tissue from the human occipital cortex from 1 field of view (FOV) of larger experiment. The data from one field of view correspond to 9 images from 6 imaging rounds (r) 3 color channels (c) and 1 z-plane (z). Each image is 988x988 (y,x)
 # EPY: END markdown
 
 # EPY: START code
 use_test_data = os.getenv("USE_TEST_DATA") is not None
-exp = data.DARTFISH(use_test_data=use_test_data)
+experiment = data.DARTFISH(use_test_data=use_test_data)
 
-stack = exp.fov().get_image(FieldOfView.PRIMARY_IMAGES)
-# EPY: END code
+imgs = experiment.fov().get_image(FieldOfView.PRIMARY_IMAGES)
 
-# EPY: START code
-print(stack.shape)
+print(imgs)
 # EPY: END code
 
+# EPY: START markdown
+### Visualize codebook
+#
+#The DARTFISH codebook maps pixel intensities across the rounds and channels to the corresponding barcodes and genes that those pixels code for. For this example dataset, the codebook specifies 96 possible barcodes.
+# EPY: END markdown
+
 # EPY: START code
-# EPY: ESCAPE %gui qt5
-display(stack)
+experiment.codebook
 # EPY: END code
 
 # EPY: START markdown
-#### Load codebook
+### Visualize raw data
+#
+#A nice way to page through all this data is to use the display command. We have commented this out for now, because it will not render in Github. Instead, we simply show an image from the first round and color channel.
 # EPY: END markdown
 
 # EPY: START code
-exp.codebook
+# %gui qt5
+# display(stack)
+
+single_plane = imgs.sel({Axes.ROUND: 0, Axes.CH: 0, Axes.ZPLANE: 0})
+single_plane = single_plane.xarray.squeeze()
+plt.figure(figsize=(7,7))
+plt.imshow(single_plane, cmap='gray', clim=list(np.percentile(single_plane, [1, 99.9])))
+plt.title('Round: 0, Chanel:0')
+plt.axis('off');
 # EPY: END code
 
 # EPY: START markdown
-#### Load copy number benchmark results
+### Filter and scale raw data before decoding into spatially resolved gene expression
+#
+#First, we normalize the intensity of each image by the maximum intensity of that image, which is equivalent to scaling by the 100th percentile value of the pixel values in each image.
 # EPY: END markdown
 
 # EPY: START code
-cnts_benchmark = pd.read_csv('https://d2nhj9g34unfro.cloudfront.net/20181005/DARTFISH/fov_001/counts.csv')
-cnts_benchmark.head()
+sc_filt = Filter.ScaleByPercentile(p=100)
+norm_imgs = sc_filt.run(imgs)
 # EPY: END code
 
 # EPY: START markdown
-#### Filter Image Stack
+#Next, for each imaging round, and each pixel location, we zero out the values across all three color channels if the magnitude of this 3 vector is below a threshold. As such, the code letters in these rounds for these pixels will be 0
```

Comment:
```suggestion
#Next, for each imaging round, and each pixel location, we zero the intensity values across all three color channels if the magnitude of this 3 vector is below a threshold. As such, the code value associated with these pixels will be the blank. This is necessary to support euclidean decoding for codebooks that include blank values. 
```

This approach feels statistically unprincipled to me. My interpretation of this method is that it's accounting for the fact that autofluorescence will bias the decoding towards non-blank channels, but this feels like it should be corrected through background removal or global intensity thresholding, not a magnitude-based zeroing. Do you recall why we didn't try those approaches when you were creating this? 

If you agree, I'd like to explore replacing this with existing filters as a follow-up.

Reply:
Many steps in many pipelines seem statistically unprincipled to me 🙃 My memory of this pipeline is fuzzy. I do vaguely remember thinking this wasn't that unreasonable of a pre-processing step. And looking at the actual data, I'm not so sure filtering would have helped much. The statistical properties of the images seem a bit odd to me.